### PR TITLE
WIP: Random Parameter Injection

### PR DESCRIPTION
**Pull request by justusschock**
## Short Description
This PR aims to inject random parameters. Therefore it consists of three parts:

- [x] The actual parameter implementation in `rising.random`
- [x] The registration of parameters in `rising.transforms.abstract`
- [x] The rewriting of all transforms to accept random parameters for certain arguments. (which arguments is still to be discussed) 

### PR Implementer
This is a small checklist for the implementation details of this PR.
If you submit a PR, please look at these points (don't worry about the `RisingTeam`
and `Reviewer` workflows, the only purpose of those is to have a compact view of
the steps)

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/PhoenixDL/rising/blob/master/CONTRIBUTING.md).

- [x] Implementation
- [ ] Docstrings & Typing
- [x] Check `__all__` sections and `__init__`
- [ ] Unittests (look at the line coverage of your tests, the goal is 100%!)
- [ ] Update notebooks & documentation if necessary
- [ ] Pass all tests
- [ ] Add the checksum of the last implementation commit to the Changelog

### RisingTeam
<details>
  <summary>RisingTeam workflow</summary>

  - [ ] Add pull request to project (optionally delete corresponding project note)
  - [ ] Assign correct label (if you don't have permission to do this, someone will do it for you. 
      Please make sure to communicate the current status of the pr.)
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)
</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>
  
  - [ ] Do all tests pass? (Unittests, NotebookTests, Documentation)
  - [ ] Does the implementation follow `rising` design conventions?
  - [ ] Are the tests useful? (!!!) Are additional tests needed? 
        Can you think of critical points which should be covered in an additional test?
  - [ ] Optional: Check coverage locally / Check tests locally if GPU is necessary to execute
</details>

### CODE REVIEW

**Comment by justusschock** [comment]
@mibaumgartner can you maybe review the current status, before I start to implement this for every transform?

File: rising/transforms/spatial.py [hunk 1]
```diff
@@ -15,9 +16,9 @@
 schduler_type = Callable[[int], Union[int, Sequence[int]]]
 
 
-class Mirror(RandomDimsTransform):
-    def __init__(self, dims: Sequence, keys: Sequence = ('data',),
-                 prob: Union[float, Sequence] = 0.5, grad: bool = False, **kwargs):
+class Mirror(BaseTransform):
+    def __init__(self, dims: Union[int, AbstractParameter, Sequence],
```
Comment (mibaumgartner) [inline rising/transforms/spatial.py:20]: we should add an integration for each case here. (mirror is a pretty easy example to test the injection with float and Sequence)

File: rising/transforms/spatial.py [hunk 2]
```diff
@@ -27,20 +28,19 @@ def __init__(self, dims: Sequence, keys: Sequence = ('data',),
...
         """
-        super().__init__(augment_fn=mirror, dims=dims, keys=keys, prob=prob, grad=grad, **kwargs)
 
+        super().__init__(augment_fn=mirror, dims=dims, keys=keys, grad=grad,
+                         property_names=('dims',), **kwargs)
 
-class Rot90(AbstractTransform):
+
+class Rot90(BaseTransform):
     def __init__(self, dims: tuple, keys: tuple = ('data',),
```
Comment (mibaumgartner) [inline rising/transforms/spatial.py:42]: Can we provide the option to pass in a `DiscreteParameter ` here?

File: rising/transforms/spatial.py [hunk 3]
```diff
@@ -62,9 +62,10 @@ def __init__(self, dims: tuple, keys: tuple = ('data',),
         --------
         :func:`torch.Tensor.rot90`
         """
-        super().__init__(grad=grad, **kwargs)
-        self.dims = dims
-        self.keys = keys
+        super().__init__(grad=grad, num_rots=DiscreteParameter((0, 1, 2, 3)),
```
Comment (mibaumgartner) [inline rising/transforms/spatial.py:65]: Should be add `num_rots` as a  `DiscreteParameter ` to our init with this default value?

File: rising/transforms/spatial.py [hunk 4]
```diff
@@ -82,33 +83,13 @@ def forward(self, **data) -> dict:
...
-        ----------
-        dims: Sequence
-            number of dimensions
-        """
-        self._dims = dims
-        self._permutations = tuple(permutations(dims, 2))
-
 
 class Resize(BaseTransform):
     def __init__(self, size: Union[int, Sequence[int]], mode: str = 'nearest',
```
Comment (mibaumgartner) [inline rising/transforms/spatial.py:95]: We should apply the same solution as in our affine rotation to the `size` parameter.

File: rising/transforms/intensity.py [hunk 5]
```diff
@@ -1,13 +1,11 @@
-import random
-from typing import Union, Sequence
-
 import torch
+from typing import Union, Sequence
 
+from rising.transforms.abstract import BaseTransform, PerSampleTransform, \
```
Comment (justusschock) [inline rising/transforms/intensity.py:4]:
```suggestion
from rising.transforms.abstract import (BaseTransform, PerSampleTransform, 
```

File: rising/transforms/intensity.py [hunk 6]
```diff
@@ -1,13 +1,11 @@
...
+    PerChannelTransform
```
Comment (justusschock) [inline rising/transforms/intensity.py:5]:
```suggestion
    PerChannelTransform)
```

File: rising/transforms/intensity.py [hunk 7]
```diff
@@ -1,13 +1,11 @@
...
 from rising.transforms.functional.intensity import norm_range, norm_min_max, norm_mean_std, \
```
Comment (justusschock) [inline rising/transforms/intensity.py:6]:
```suggestion
from rising.transforms.functional.intensity import (norm_range, norm_min_max, norm_mean_std, 
```

File: rising/transforms/intensity.py [hunk 8]
```diff
@@ -1,13 +1,11 @@
...
-    norm_zero_mean_unit_std, add_noise, gamma_correction, add_value, scale_by_value
-from rising.utils import check_scalar
-from .abstract import BaseTransform, PerSampleTransform, AbstractTransform, \
-    PerChannelTransform, RandomProcess
+    norm_zero_mean_unit_std, add_noise, gamma_correction, add_value, scale_by_value, clamp
```
Comment (justusschock) [inline rising/transforms/intensity.py:7]:
```suggestion
    norm_zero_mean_unit_std, add_noise, gamma_correction, add_value, scale_by_value, clamp)
```

File: rising/utils/checktype.py [hunk 9]
```diff
@@ -9,4 +11,9 @@ def check_scalar(x):
     Returns:
         True if input is scalar
     """
-    return isinstance(x, (int, float))
+    if isinstance(x, (int, float)):
```
Comment (justusschock) [inline rising/utils/checktype.py:14]:
```suggestion
    return isinstance(x, (int, float)) or (is instance(x, torch.Tensor) and x.numel() == 1)
```

File: rising/utils/checktype.py [hunk 10]
```diff
@@ -9,4 +11,9 @@ def check_scalar(x):
...
+        return True
+    elif torch.is_tensor(x):
```
Comment (justusschock) [inline rising/utils/checktype.py:16]: don't use this :) instead use `is instance(x, torch.Tensor)`, which is internally called by `torch.is_tensor`. `torch.is_tensor` is a legacy API from the time when tensors were no native python class

File: rising/utils/shape.py [hunk 11]
```diff
@@ -0,0 +1,18 @@
+import torch
+from typing import Union, Sequence
+
+
+def reshape(value: Union[list, torch.Tensor],
```
Comment (justusschock) [inline rising/utils/shape.py:5]: docstrings? :)

File: rising/utils/shape.py [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+            size: Union[Sequence, torch.Size]) -> Union[torch.Tensor, list]:
+    if isinstance(value, torch.Tensor):
+        return value.view(size)
+
+    else:
+        return reshape_list(value, size)
+
+
+def reshape_list(flat_list: list, size: Union[torch.Size, tuple]) -> list:
```
Comment (justusschock) [inline rising/utils/shape.py:14]: also docstrings? :)

File: tests/transforms/test_affine.py [hunk 13]
```diff
@@ -1,8 +1,7 @@
 import unittest
 from rising.transforms.affine import Affine, StackedAffine, Translate, Rotate, \
```
Comment (justusschock) [inline tests/transforms/test_affine.py:2]:
```suggestion
from rising.transforms.affine import (Affine, StackedAffine, Translate, Rotate,
```

File: tests/transforms/test_affine.py [hunk 14]
```diff
@@ -1,8 +1,7 @@
...
-    Scale, Resize
+    Scale, Resize, BaseAffine
```
Comment (justusschock) [inline tests/transforms/test_affine.py:3]:
```suggestion
    Scale, Resize, BaseAffine)
```

File: rising/random/abstract.py [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+    'AbstractParameter'
+]
+
+
+class AbstractParameter(torch.nn.Module):
+    """
+    Abstract Parameter class to inject randomness to transforms
+    """
+
+    def __init__(self, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (mibaumgartner) [reply rising/random/abstract.py:17]: You are right, forgot to change it (was looking for a bug in `sample`)

File: rising/utils/checktype.py (hunk 10 above)
Comment (mibaumgartner) [reply rising/utils/checktype.py:16]: 👍 Learned something new :)

File: rising/random/continuous.py [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+from rising.random.abstract import AbstractParameter
+
+__all__ = [
+    'ContinuousParameter',
+    'NormalParameter',
+    'UniformParameter'
+]
+
+
+class ContinuousParameter(AbstractParameter):
```
[the remark this answers is not in the document]
Comment (mibaumgartner) [reply rising/random/continuous.py:15]: I think we will postpone this to the next release due to time constraints 👍 There are also some other things we can consider (but we keep the same external API). We should probably discuss this in an Issue :)
